File: docs/quantum_spin_lang/berry_phase.R
# Experiment 1: Berry Phase = Pythagorean Comma
# Start f=440, multiply by 3/2 twelve times, normalize to octave.
# Discrepancy = 23.46 cents (Pythagorean comma).

cents <- function(ratio) 1200 * log2(ratio)

main <- function() {
  base_freq <- 440
  fifth <- 3/2
  n_steps <- 12
  
  # Compute 12 perfect fifths
  frequencies <- numeric(n_steps)
  f <- base_freq
  for (i in 1:n_steps) {
    f <- f * fifth
    frequencies[i] <- f
  }
  
  # Normalize to base octave
  normalized <- sapply(frequencies, function(freq) {
    ratio <- freq / base_freq
    octaves <- floor(log2(ratio))
    freq / (2^octaves)
  })
  
  # Pythagorean comma
  pythagorean_comma_ratio <- (3/2)^12 / 2^7
  pythagorean_comma_cents <- cents(pythagorean_comma_ratio)
  
  # Build results
  steps <- lapply(1:n_steps, function(i) {
    list(step = i, freq = frequencies[i], normalized = normalized[i])
  })
  
  sorted <- sort(normalized)
  
  result <- list(
    experiment = "Berry Phase = Pythagorean Comma",
    base_freq = base_freq,
    fifth_ratio = 1.5,
    steps = n_steps,
    frequencies = steps,
    sorted_normalized = round(sorted, 6),
    pythagorean_comma_ratio = round(pythagorean_comma_ratio, 10),
    pythagorean_comma_cents = round(pythagorean_comma_cents, 4),
    expected_cents = 23.46,
    match = abs(pythagorean_comma_cents - 23.46) < 0.1
  )
  
  cat(toJSON(result))
}

# Minimal JSON serializer (base R)
toJSON <- function(x, indent = 0) {
  pad <- paste(rep("  ", indent), collapse = "")
  pad1 <- paste(rep("  ", indent + 1), collapse = "")
  
  if (is.list(x)) {
    if (!is.null(names(x))) {
      # Named list -> object
      entries <- sapply(names(x), function(nm) {
        val <- toJSON(x[[nm]], indent + 1)
        paste0(pad1, '"', nm, '": ', val)
      })
      paste0("{\n", paste(entries, collapse = ",\n"), "\n", pad, "}")
    } else {
      # Unnamed list -> array of objects
      entries <- sapply(x, function(item) {
        paste0(pad1, toJSON(item, indent + 1))
      })
      paste0("[\n", paste(entries, collapse = ",\n"), "\n", pad, "]")
    }
  } else if (is.numeric(x)) {
    if (length(x) == 1) {
      if (x == round(x) && abs(x) < 1e15) as.character(x) else sprintf("%.6f", x)
    } else {
      vals <- sapply(x, function(v) if (v == round(v) && abs(v) < 1e15) as.character(v) else sprintf("%.6f", v))
      paste0("[", paste(vals, collapse = ", "), "]")
    }
  } else if (is.logical(x)) {
    tolower(as.character(x))
  } else if (is.character(x)) {
    paste0('"', x, '"')
  } else {
    paste0('"', as.character(x), '"')
  }
}

# Plot
make_plot <- function() {
  base_freq <- 440
  fifth <- 3/2
  
  frequencies <- numeric(12)
  f <- base_freq
  for (i in 1:12) { f <- f * fifth; frequencies[i] <- f }
  
  normalized <- sapply(frequencies, function(freq) {
    ratio <- freq / base_freq
    octaves <- floor(log2(ratio))
    freq / (2^octaves)
  })
  
  sorted <- sort(normalized)
  
  png("/home/phoenix/.openclaw/workspace/docs/quantum_spin_lang/berry_phase_plot.png",
      width = 800, height = 600, res = 120)
  
  # Circle of fifths as helix
  angles <- seq(0, 2*pi - pi/6, length.out = 12)
  expected <- base_freq * (2^(0:11/12))
  
  plot(normalized, angles, pch = 19, col = "blue",
       xlab = "Frequency (Hz)", ylab = "Phase angle (rad)",
       main = "Berry Phase: Circle of Fifths\nDiscrepancy = Pythagorean Comma (23.46 cents)")
  points(expected, angles, pch = 17, col = "red")
  legend("topleft", legend = c("Perfect fifths", "Equal temperament"),
         pch = c(19, 17), col = c("blue", "red"))
  
  dev.off()
}

make_plot()
main()

File: docs/quantum_spin_lang/rust/src/berry_phase.rs
//! Experiment 1: Berry Phase = Pythagorean Comma
//! Start f=440, multiply by 3/2 twelve times, normalize to octave.
//! Discrepancy = 23.46 cents (Pythagorean comma).

use std::f64::consts::LOG2_E;

const LN2: f64 = std::f64::consts::LN_2;

fn cents(ratio: f64) -> f64 {
    1200.0 * ratio.log2()
}

fn main() {
    let base_freq = 440.0;
    let fifth = 3.0_f64 / 2.0;
    let n_steps = 12;

    // Compute 12 perfect fifths
    let mut frequencies: Vec<f64> = Vec::with_capacity(n_steps);
    let mut f = base_freq;
    for i in 0..n_steps {
        f *= fifth;
        frequencies.push(f);
    }

    // Normalize all to base octave [base_freq, 2*base_freq)
    let normalized: Vec<f64> = frequencies
        .iter()
        .map(|&freq| {
            let ratio = freq / base_freq;
            let octaves = ratio.log2().floor() as i32;
            freq / (2.0_f64.powi(octaves))
        })
        .collect();

    // The 12th fifth should land near 2*base_freq (7 octaves up)
    // (3/2)^12 = 531441/4096 ≈ 129.746
    // 2^7 = 128
    // Ratio: 531441/4096 / 128 = 531441/524288
    let pythagorean_comma_ratio = (3.0_f64 / 2.0).powi(12) / 2.0_f64.powi(7);
    let pythagorean_comma_cents = cents(pythagorean_comma_ratio);

    // Sort normalized frequencies to see the circle of fifths scale
    let mut sorted = normalized.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());

    // Build JSON output
    let freqs_json: Vec<String> = frequencies
        .iter()
        .enumerate()
        .map(|(i, f)| format!("{{\"step\": {}, \"freq\": {:.6}, \"normalized\": {:.6}}}", i + 1, f, normalized[i]))
        .map(|s| s)
        .collect();

    let sorted_json: Vec<String> = sorted
        .iter()
        .map(|f| format!("{:.6}", f))
        .collect();

    let json = format!(
        "{{\n  \"experiment\": \"Berry Phase = Pythagorean Comma\",\n  \"base_freq\": {:.1},\n  \"fifth_ratio\": 1.5,\n  \"steps\": {},\n  \"frequencies\": [\n    {}\n  ],\n  \"sorted_normalized\": [{}],\n  \"pythagorean_comma_ratio\": {:.10},\n  \"pythagorean_comma_cents\": {:.4},\n  \"expected_cents\": 23.46,\n  \"match\": {}\n}}",
        base_freq,
        n_steps,
        freqs_json.join(",\n    "),
        sorted_json.join(", "),
        pythagorean_comma_ratio,
        pythagorean_comma_cents,
        (pythagorean_comma_cents - 23.46).abs() < 0.1
    );

    println!("{}", json);
}
